Code/Framework/AzCore/AzCore/DOM/DomBackend.cpp

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include <AzCore/DOM/DomBackend.h>
+
+#include <AzCore/Utils/Utils.h>
+#include <AzCore/IO/ByteContainerStream.h>
+#include "DomBackend.h"
+
+namespace AZ::DOM
+{
+    Visitor::Result Backend::ReadFromPath(AZ::IO::PathView pathName, Visitor* visitor, size_t maxFileSize)
```

Comment:
I'm not convinced that the generic dom should concern itself with reading/writing data. Imho it just muddies the waters as it's main purpose is to manage a DOM, not file management. For comparison, the Json Serialization doesn't do any file handling, nor does RapidJSON. It's fine to have utility functions to deal with this, but those already exist such as the utility functions to read a JSON document into RapidJSON. Imho, this is ultimately yet another variation of the same code and only adds bloat to the code base.

Reply:
To add on to Ronald's point, the SettingsRegistry doesn't provide writing to a file or string .
There are utility functions within the SettingsRegistryMergeUtils for dumping to a [stream](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/Settings/SettingsRegistryMergeUtils.cpp#L1153-L1315).

Even just making assumptions as to what to use to open a file, can cause issues problems going forward such as reading and writing using the FileIOSystem.

For example the [AZ::Utils::ReadFile](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/Utils/Utils.cpp#L147-L176) that is being used  is using the FileIOStream under the hood.
This will not work if the FileIOBase instance hasn't been created yet.
This function is actually adding a dependency to the Backend on the FileIO system which might not be available, where something like the SystemFile API can be used.

I believe if anything we should only have functions that work with the GenericStream API.
We have [ByteContainerStream<T>](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/IO/ByteContainerStream.h#L36-L38) that can be used to write to `AZStd::string`, `AZStd::fixed_string<Num>`, `AZStd::vector<T>`, `AZStd::fixed_vector<T, Num>`, etc...

We have [SystemFileStream](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/IO/GenericStreams.cpp#L141-L144) for using the SystemFile API and we have [FileIOStream](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/IO/FileIO.h#L267-L271) that uses the FileIO API.

Even out of all those streams classes I listed, the only one I believe should ever be used is the `ByteContainerStream` as it has been proven to have been significantly faster when it comes to writing to an in memory buffer and then writing to disk.
We saw massive save file improvements with slices by writing to a ByteContainerStream first(see CL 427497 in a certain source control tool)
> Optimize saving slice assets to disk - now 30x quicker
>
> - Instead of writing straight to FileStream, we now write to a MemoryStream first and then from the MemoryStream to FileStream. Takes the time to write a 2500 loose entity slice to disk from 84 seconds down to 2.75 seconds (and that 2.75 seconds includes both writing to memory and then writing from memory to disk).
>
> - We previously saved slice ObjectStream as XML to FileStream directly. rapidxml library's print() function (which writes its in-memory representation of the XML doc to a bytestream) writes a character at a time to FileStream, each character going through a significant conditional-ridden callstack (including into CryPak land). Writing a string of only a few characters was taking hundreds of microseconds. We now write straight to a byte stream in memory, so that the byte-by-byte writing done by rapidxml can be rapid, and then we just do one big memory -> file write, so all of that overhead is reduced significantly.

Reply:
I've removed the file IO part of the API in favor of a stream API, though I'd still like to kick around where we'd like some generic "read from a file and do it efficiently (i.e. read into memory stream then parse)" for generic DOM operations, it's not critical for anything I'm doing ATM.